File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.py
import sys
input = sys.stdin.readline
from copy import deepcopy

n = int(input())
board = [list(map(int, input().split())) for _ in range(n)]

def up(graph):
    for j in range(n):
        pointer = 0
        for i in range(1,n):
            if graph[i][j]:
                tmp = graph[i][j]
                graph[i][j] = 0

                if graph[pointer][j] == 0:
                    graph[pointer][j] = tmp
                
                elif graph[pointer][j] == tmp:
                    graph[pointer][j] *= 2
                    pointer += 1

                else:
                    pointer += 1
                    graph[pointer][j] = tmp
    return graph

def down(graph):
    for j in range(n):
        pointer = n-1
        for i in range(n-2,-1,-1):
            if graph[i][j]:
                tmp = graph[i][j]
                graph[i][j] = 0

                if graph[pointer][j] == 0:
                    graph[pointer][j] = tmp
                
                elif graph[pointer][j] == tmp:
                    graph[pointer][j] *= 2
                    pointer -= 1

                else:
                    pointer -= 1
                    graph[pointer][j] = tmp
    return graph
            
def left(graph):
    for i in range(n):
        pointer = 0
        for j in range(1,n):
            if graph[i][j]:
                tmp = graph[i][j]
                graph[i][j] = 0

                if graph[i][pointer] == 0:
                    graph[i][pointer] = tmp

                elif graph[i][pointer]  == tmp:
                    graph[i][pointer] *= 2
                    pointer += 1

                else:
                    pointer += 1
                    graph[i][pointer]= tmp                    
    return graph

def right(graph):
    for i in range(n):
        pointer = n - 1
        for j in range(n - 2, -1, -1):
            if graph[i][j]:
                tmp = graph[i][j]
                graph[i][j] = 0
                if graph[i][pointer] == 0:
                    graph[i][pointer] = tmp
                elif graph[i][pointer]  == tmp:
                    graph[i][pointer] *= 2
                    pointer -= 1
                else:
                    pointer -= 1
                    graph[i][pointer] = tmp
    return graph

ans = 0
def dfs(graph, cnt):
    global ans

    if cnt == 5:
        ans = max(ans, max(map(max,graph)))
        return
    
    dfs(up(deepcopy(graph)), cnt + 1)
    dfs(down(deepcopy(graph)), cnt + 1)
    dfs(left(deepcopy(graph)), cnt + 1)
    dfs(right(deepcopy(graph)), cnt + 1)

dfs(board,0)
print(ans)

File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	static int n;
	static int[][] arr;
	static int ans = 0;

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st;
		n = Integer.parseInt(br.readLine());
		arr = new int[n][n];

		// 배열 값 갱신
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < n; j++) {
				arr[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		dfs(arr, 0); // 백트래킹 시작

		sb.append(ans);
		System.out.println(sb.toString());
	}

	static void dfs(int[][] arr, int cnt) {
		if (cnt == 5) {
			int value = findMax(arr);
			if (ans < value) {
				ans = value;
			}
			return;
		}

		dfs(goLeft(arr), cnt + 1);
		dfs(goRight(arr), cnt + 1);
		dfs(goUp(arr), cnt + 1);
		dfs(goDown(arr), cnt + 1);
	}

	static int[][] goLeft(int[][] arr) {
		int[][] newArr = cloneArr(arr);

		for (int i = 0; i < n; i++) {
			int mainPointer = 0;
			for (int j = mainPointer + 1; j < n; j++) {
				if (newArr[i][j] != 0) {
					if (newArr[i][mainPointer] == 0) {
						newArr[i][mainPointer] = newArr[i][j];
						newArr[i][j] = 0;
					} else if (newArr[i][mainPointer] == newArr[i][j]) {
						newArr[i][mainPointer] *= 2;
						newArr[i][j] = 0;
						mainPointer++;

					} else {
						mainPointer++;
						int tmp = newArr[i][mainPointer];
						newArr[i][mainPointer] = newArr[i][j];
						newArr[i][j] = tmp;
					}
				}
			}
		}
		return newArr;
	}

	static int[][] goRight(int[][] arr) {
		int[][] newArr = cloneArr(arr);

		for (int i = n - 1; i > -1; i--) {
			int mainPointer = n - 1;
			for (int j = mainPointer - 1; j > -1; j--) {
				if (newArr[i][j] != 0) {
					if (newArr[i][mainPointer] == 0) {
						newArr[i][mainPointer] = newArr[i][j];
						newArr[i][j] = 0;
					} else if (newArr[i][mainPointer] == newArr[i][j]) {
						newArr[i][mainPointer] *= 2;
						newArr[i][j] = 0;
						mainPointer--;
					} else {
						mainPointer--;
						int tmp = newArr[i][mainPointer];
						newArr[i][mainPointer] = newArr[i][j];
						newArr[i][j] = tmp;
					}
				}
			}
		}

		return newArr;
	}

	static int[][] goUp(int[][] arr) {
		int[][] newArr = cloneArr(arr);

		for (int j = 0; j < n; j++) {
			int mainPointer = 0;
			for (int i = mainPointer + 1; i < n; i++) {
				if (newArr[i][j] != 0) {
					if (newArr[mainPointer][j] == 0) {
						newArr[mainPointer][j] = newArr[i][j];
						newArr[i][j] = 0;
					} else if (newArr[mainPointer][j] == newArr[i][j]) {
						newArr[mainPointer][j] *= 2;
						newArr[i][j] = 0;
						mainPointer++;
					} else {
						mainPointer++;
						int tmp = newArr[mainPointer][j];
						newArr[mainPointer][j] = newArr[i][j];
						newArr[i][j] = tmp;
					}
				}
			}
		}

		return newArr;
	}

	static int[][] goDown(int[][] arr) {
		int[][] newArr = cloneArr(arr);

		for (int j = n - 1; j > -1; j--) {
			int mainPointer = n - 1;
			for (int i = mainPointer - 1; i > -1; i--) {
				if (newArr[i][j] != 0) {
					if (newArr[mainPointer][j] == 0) {
						newArr[mainPointer][j] = newArr[i][j];
						newArr[i][j] = 0;
					} else if (newArr[mainPointer][j] == newArr[i][j]) {
						newArr[mainPointer][j] *= 2;
						newArr[i][j] = 0;
						mainPointer--;
					} else {
						mainPointer--;
						int tmp = newArr[mainPointer][j];
						newArr[mainPointer][j] = newArr[i][j];
						newArr[i][j] = tmp;
					}
				}
			}
		}

		return newArr;
	}

	// 2차원 배열 복사
	static int[][] cloneArr(int[][] arr) {
		int[][] copy = new int[arr.length][arr[0].length];

		for (int i = 0; i < arr.length; i++) {
			copy[i] = arr[i].clone();
		}
		return copy;
	}

	// 최대 값 찾기
	static int findMax(int[][] arr) {
		int max = 0;
		for (int[] line : arr) {
			for (int value : line) {
				if (value > max)
					max = value;
			}
		}
		return max;
	}
}